src/main/java/bio/terra/service/job/StairwayExceptionFieldsFactory.java

```diff
@@ -0,0 +1,103 @@
+package bio.terra.service.job;
+
+import bio.terra.common.exception.ErrorReportException;
+import bio.terra.service.job.exception.ExceptionSerializerException;
+import bio.terra.service.job.exception.JobExecutionException;
+import bio.terra.service.job.exception.StepExecutionException;
+import bio.terra.stairway.Step;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.apache.commons.lang3.StringUtils;
+
+public class StairwayExceptionFieldsFactory {
+
+  private static String CONTACT_TEAM_MESSAGE = "Please contact the TDR team for help.";
+
+  public static StairwayExceptionFields fieldsFromException(Exception ex) {
+    Optional<String> stepName = getStepNameFromStairwayException(ex);
+
+    Optional<String> fieldExceptionClassName = Optional.empty();
+    if (ErrorReportException.class.isAssignableFrom(ex.getClass())) {
+      fieldExceptionClassName = Optional.of(ex.getClass().getName());
+    }
+
+    if (stepName.isEmpty()) {
+      return new StairwayExceptionFields()
+          .setClassName(fieldExceptionClassName.orElse(JobExecutionException.class.getName()))
+          .setDataRepoException(true)
+          .setMessage(getJobExceptionMessage(ex))
+          .setErrorDetails(getJobExceptionDetails(ex));
+    }
+
+    return new StairwayExceptionFields()
+        .setClassName(fieldExceptionClassName.orElse(StepExecutionException.class.getName()))
+        .setDataRepoException(true)
+        .setMessage(getStepExceptionMessage(stepName.get(), ex))
+        .setErrorDetails(getStepExceptionDetails(ex));
+  }
+
+  private static Optional<String> getStepNameFromStairwayException(Exception ex) {
+    String failedStepName = null;
+    for (StackTraceElement ste : ex.getStackTrace()) {
+      try {
+        Class<?> cls = Class.forName(ste.getClassName());
+        if (Step.class.isAssignableFrom(cls)) {
+          // Java 17 will make this casting unnecessary
```

Comment:
I don't see a cast here. Do you mean the use of reflection to get the class type for the stack element?